### PR TITLE
[backport v1.14] do some necessary backport for ARC

### PR DESCRIPTION
**Pull request by IRISZZW**
1. fix the bug where regular irq comes out in user mode
2. fix the bug where buffer validate should be atomic
3. add handling for r25, r30
4. Preserve STATUS32 flags while setting AE flag

### CODE REVIEW

File: arch/arc/core/isr_wrapper.S [hunk 1]
```diff
@@ -337,6 +337,13 @@ _skip_sys_power_save_idle_exit:
 SECTION_FUNC(TEXT, _isr_demux)
 	push_s r3
 
+/* according to ARCv2 ISA, r25, r30, r58, r59 are caller-saved
+ * scratch registers, possibly used by interrupt handlers
+ */
+	push r25
+	push r30
+	push r58
```
Comment (ruuddw) [inline arch/arc/core/isr_wrapper.S:345]: for the accumulator registers (r58/r59), the CONFIG_ARC_HAS_ACCL_REGS should be used/backported as well. Only DSP-flavors of ARC have these.

File: arch/arc/core/mpu/arc_mpu_v3_internal.h [hunk 2]
```diff
@@ -596,13 +596,15 @@ int arc_core_mpu_buffer_validate(void *addr, size_t size, int write)
...
+			r_index = 0;
 		} else {
-			return -EPERM;
+			r_index = -EPERM;
 		}
+	} else {
+		r_index = -EPERM;
 	}
 
-	return -EPERM;
```
Comment (ruuddw) [inline arch/arc/core/mpu/arc_mpu_v3_internal.h:605]: where is the unlock?

**Review by ruuddw — CHANGES_REQUESTED** [review]
Looks like some crucial parts were missed in the backport, see comments.

File: arch/arc/core/isr_wrapper.S (hunk 1 above)
Comment (andrewboie) [reply arch/arc/core/isr_wrapper.S:345]: @ruuddw this is a backport PR, is there a patch in master for this one?

Comment (IRISZZW) [reply arch/arc/core/isr_wrapper.S:345]: ARC FPU is not support in 1.14 (supported in PR: #16921), so we can remove the save and restore of this two registers.

File: arch/arc/core/mpu/arc_mpu_v3_internal.h (hunk 2 above)
Comment (IRISZZW) [reply arch/arc/core/mpu/arc_mpu_v3_internal.h:605]: so stupid mistake, fixed.

**Review by ruuddw — APPROVED** [review]

**Comment by ruuddw** [comment]
Looks ok to me now. @IRISZZW can you confirm regression tests pass, also on hardware boards?

**Pull request merged**